File: Projects/Offensive/CSRF checker/csrf_checker.py
#!/usr/bin/env python3
"""
csrf_checker.py
Usage: python3 csrf_checker.py <url>
Example: python3 csrf_checker.py https://example.com/login
Notes: Safe, read-only scanner. Does not submit forms or perform harmful actions.
Dependencies: pip install requests beautifulsoup4
"""
import sys
import requests
from urllib.parse import urljoin, urlparse
from bs4 import BeautifulSoup

COMMON_CSRF_NAMES = {
    "csrf", "csrf_token", "csrf-token", "authenticity_token",
    "_csrf", "__requestverificationtoken", "token", "xsrf-token", "anti_csrf"
}
STATE_GET_KEYWORDS = ("delete", "remove", "logout", "signout", "transfer", "withdraw", "confirm")

def fetch(url):
    try:
        r = requests.get(url, timeout=10)
        return r
    except Exception as e:
        print(f"[!] Error fetching {url}: {e}")
        return None

def cookie_checks(response):
    cookies = []
    raw = response.headers.get("Set-Cookie")
    if not raw:
        return cookies
    # requests exposes cookie attributes partially; parse header for attributes
    for header in response.headers.get_all("Set-Cookie") if hasattr(response.headers, "get_all") else [raw]:
        cookies.append(header)
    return cookies

def has_csrf_token(inputs):
    for inp in inputs:
        name = (inp.get("name") or "").lower()
        _id = (inp.get("id") or "").lower()
        typ = (inp.get("type") or "").lower()
        # hidden inputs with common csrf-like names
        if typ == "hidden" and (name in COMMON_CSRF_NAMES or _id in COMMON_CSRF_NAMES):
            return True, name or _id
        # some tokens are not hidden but named explicitly
        if name in COMMON_CSRF_NAMES or _id in COMMON_CSRF_NAMES:
            return True, name or _id
    return False, None

def analyze_forms(soup, base_url):
    findings = []
    for form in soup.find_all("form"):
        action = form.get("action") or ""
        action = urljoin(base_url, action)
        method = (form.get("method") or "get").lower()
        inputs = form.find_all("input")
        token_found, token_name = has_csrf_token(inputs)
        # Heuristic: consider forms with POST and no token as risky
        risk = None
        if method == "post" and not token_found:
            risk = "POST form without detectable CSRF token"
        # mark state-changing GET forms/links
        if method == "get":
            for kw in STATE_GET_KEYWORDS:
                if kw in action.lower():
                    if risk:
                        risk += f"; GET action contains '{kw}'"
                    else:
                        risk = f"GET action contains state keyword '{kw}'"
        findings.append({
            "action": action,
            "method": method,
            "inputs_count": len(inputs),
            "token_found": token_found,
            "token_name": token_name,
            "risk": risk
        })
    return findings

def analyze_links(soup, base_url):
    findings = []
    for a in soup.find_all("a", href=True):
        href = urljoin(base_url, a['href'])
        for kw in STATE_GET_KEYWORDS:
            if kw in href.lower():
                findings.append({"url": href, "keyword": kw})
                break
    return findings

def report(url):
    print(f"Scanning (read-only): {url}\n")
    resp = fetch(url)
    if resp is None:
        return
    print(f"HTTP {resp.status_code} {resp.reason}")
    # cookies
    raw_cookies = cookie_checks(resp)
    if raw_cookies:
        print("\nSet-Cookie headers found:")
        for c in raw_cookies:
            print("  " + c)
            if "samesite" not in c.lower():
                print("    -> Missing SameSite attribute (recommend 'Lax' or 'Strict').")
    else:
        print("\nNo Set-Cookie header found on initial GET.")

    soup = BeautifulSoup(resp.text, "html.parser")
    forms = analyze_forms(soup, url)
    if forms:
        print(f"\nFound {len(forms)} form(s):")
        for i, f in enumerate(forms, 1):
            print(f" {i}. action={f['action']} method={f['method'].upper()} inputs={f['inputs_count']}")
            if f['token_found']:
                print(f"    - CSRF token detected (field: {f['token_name']})")
            else:
                print("    - NO detectable CSRF token.")
            if f['risk']:
                print(f"    - RISK: {f['risk']}")
    else:
        print("\nNo forms detected on the page.")

    links = analyze_links(soup, url)
    if links:
        print(f"\nPotential state-changing GET links ({len(links)}):")
        for l in links:
            print(f"  - {l['url']} (contains '{l['keyword']}')")
    else:
        print("\nNo obvious state-changing GET links detected heuristically.")

    print("\nRecommendations:")
    print(" - Add per-request anti-CSRF tokens to forms and verify server-side.")
    print(" - Use SameSite=Lax or Strict on session cookies where practical.")
    print(" - Avoid performing state changes via GET requests.")
    print(" - Verify Referer/Origin server-side as an additional check for sensitive endpoints.")
    print("\nNote: this is a heuristic, read-only tool. False positives/negatives are possible.")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python3 csrf_checker.py <url>")
        sys.exit(1)
    target = sys.argv[1]
    report(target)

File: Projects/Offensive/CSRF checker/csrf_checker.c
/* csrf_checker.c
   Compile: gcc -o csrf_checker csrf_checker.c -lcurl
   Usage: ./csrf_checker <url>
   Notes: Read-only scanner. Requires libcurl development package.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>

struct mem {
    char *data;
    size_t length;
};

static size_t write_cb(void *ptr, size_t size, size_t nmemb, void *userp) {
    size_t realsz = size * nmemb;
    struct mem *mem = (struct mem*)userp;
    char *tmp = realloc(mem->data, mem->length + realsz + 1);
    if (!tmp) return 0;
    mem->data = tmp;
    memcpy(&(mem->data[mem->length]), ptr, realsz);
    mem->length += realsz;
    mem->data[mem->length] = 0;
    return realsz;
}

int contains_case_insensitive(const char *hay, const char *needle) {
    if (!hay || !needle) return 0;
    size_t hn = strlen(hay), nn = strlen(needle);
    for (size_t i=0;i+nn<=hn;i++) {
        size_t j=0;
        for (; j<nn; j++) {
            char a = hay[i+j], b = needle[j];
            if (a >= 'A' && a <= 'Z') a += 32;
            if (b >= 'A' && b <= 'Z') b += 32;
            if (a != b) break;
        }
        if (j==nn) return 1;
    }
    return 0;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <url>\n", argv[0]);
        return 1;
    }
    const char *url = argv[1];
    CURL *curl = curl_easy_init();
    if (!curl) {
        fprintf(stderr, "libcurl init failed\n");
        return 1;
    }
    struct mem chunk = {0};
    struct curl_slist *hdrs = NULL;
    char header_buf[1024];

    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &chunk);

    // collect headers manually for Set-Cookie
    curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, write_cb);
    curl_easy_setopt(curl, CURLOPT_HEADERDATA, &chunk);

    CURLcode res = curl_easy_perform(curl);
    if (res != CURLE_OK) {
        fprintf(stderr, "Request failed: %s\n", curl_easy_strerror(res));
        curl_easy_cleanup(curl);
        free(chunk.data);
        return 1;
    }

    long http_code = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
    printf("HTTP %ld\n", http_code);

    // Look for Set-Cookie and SameSite
    if (chunk.length > 0 && contains_case_insensitive(chunk.data, "Set-Cookie:")) {
        printf("\nSet-Cookie header(s) present (raw):\n");
        const char *p = chunk.data;
        while ((p = strstr(p, "Set-Cookie:")) != NULL) {
            const char *eol = strstr(p, "\r\n");
            if (!eol) eol = p + strlen(p);
            size_t len = eol - p;
            if (len > 0 && len < sizeof(header_buf)) {
                strncpy(header_buf, p, len);
                header_buf[len] = 0;
                printf("  %s\n", header_buf);
                if (!contains_case_insensitive(header_buf, "samesite")) {
                    printf("    -> Missing SameSite attribute (recommend Lax or Strict)\n");
                }
            }
            p = eol ? eol+2 : p+9;
        }
    } else {
        printf("\nNo Set-Cookie headers detected in response headers.\n");
    }

    // Simple form checks (very heuristic)
    const char *html = chunk.data ? chunk.data : "";
    int form_count = 0, risk_forms = 0;
    const char *pos = html;
    while ((pos = strcasestr(pos, "<form")) != NULL) {
        form_count++;
        const char *endform = strcasestr(pos, "</form>");
        size_t blocklen = endform ? (endform - pos) : strlen(pos);
        char *block = malloc(blocklen + 1);
        strncpy(block, pos, blocklen);
        block[blocklen] = 0;

        // detect method
        int is_post = contains_case_insensitive(block, "method=\"post\"") || contains_case_insensitive(block, "method='post'");

        // detect hidden inputs with common names
        const char *names[] = {
            "csrf", "csrf_token","csrf-token","authenticity_token",
            "_csrf","__requestverificationtoken","token","xsrf-token","anti_csrf", NULL
        };
        int token_found = 0;
        for (int i=0; names[i]; ++i) {
            if (contains_case_insensitive(block, names[i])) { token_found = 1; break; }
        }
        if (is_post && !token_found) {
            printf("\nForm %d: POST form without detectable CSRF token (heuristic)\n", form_count);
            risk_forms++;
        }
        free(block);
        pos = endform ? endform + 7 : pos + 5;
    }

    printf("\nForms found (heuristic): %d\n", form_count);
    if (risk_forms) {
        printf("Potentially risky POST forms without CSRF token: %d\n", risk_forms);
    } else {
        printf("No obviously risky POST forms detected by this heuristic.\n");
    }

    printf("\nRecommendations:\n - Use per-request server-side CSRF tokens in forms.\n - Set SameSite cookie attribute (Lax/Strict) where suitable.\n - Avoid state-changing GET endpoints.\n - Implement server-side Referer/Origin checks for sensitive endpoints.\n");

    curl_easy_cleanup(curl);
    free(chunk.data);
    return 0;
}
